File: src/cz/cuni/lf1/lge/ThunderSTORM/filters/ui/CompoundWaveletFilterUI.java
package cz.cuni.lf1.lge.ThunderSTORM.filters.ui;

import cz.cuni.lf1.lge.ThunderSTORM.filters.CompoundWaveletFilter;
import cz.cuni.lf1.lge.ThunderSTORM.filters.IFilter;
import cz.cuni.lf1.lge.ThunderSTORM.util.GridBagHelper;
import cz.cuni.lf1.lge.thunderstorm.util.macroui.ParameterKey;
import cz.cuni.lf1.lge.thunderstorm.util.macroui.validators.DoubleValidatorFactory;
import cz.cuni.lf1.lge.thunderstorm.util.macroui.validators.IntegerValidatorFactory;
import java.awt.GridBagLayout;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JTextField;

public class CompoundWaveletFilterUI extends IFilterUI {

    private final String name = "Wavelet filter (B-Spline)";
    private transient ParameterKey.Integer size;
    private transient ParameterKey.Integer order;
    private transient ParameterKey.Double scale;

    public CompoundWaveletFilterUI() {
        size = parameters.createIntField("size", IntegerValidatorFactory.positiveNonZero(), 5);
        scale = parameters.createDoubleField("scale", DoubleValidatorFactory.positiveNonZero(), 2.0);
        order = parameters.createIntField("order", IntegerValidatorFactory.positiveNonZero(), 3);
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    protected String getPreferencesPrefix() {
        return super.getPreferencesPrefix() + ".wave";
    }

    @Override
    public JPanel getOptionsPanel() {
        JTextField orderTextField = new JTextField("", 20);
        JTextField scaleTextField = new JTextField("", 20);
        JTextField sizeTextField = new JTextField("", 20);
        parameters.registerComponent(order, orderTextField);
        parameters.registerComponent(scale, scaleTextField);
        parameters.registerComponent(size, sizeTextField);
        //
        JPanel panel = new JPanel(new GridBagLayout());
        panel.add(new JLabel("B-Spline order: "), GridBagHelper.leftCol());
        panel.add(orderTextField, GridBagHelper.rightCol());
        panel.add(new JLabel("B-Spline scale: "), GridBagHelper.leftCol());
        panel.add(scaleTextField, GridBagHelper.rightCol());
        panel.add(new JLabel("Kernel size [px]: "), GridBagHelper.leftCol());
        panel.add(sizeTextField, GridBagHelper.rightCol());
        parameters.loadPrefs();
        return panel;
    }

    @Override
    public IFilter getImplementation() {
        return new CompoundWaveletFilter(order.getValue(), scale.getValue(), size.getValue());
    }
}

File: src/main/java/cz/cuni/lf1/lge/ThunderSTORM/filters/ui/CompoundWaveletFilterUI.kt
package cz.cuni.lf1.lge.ThunderSTORM.filters.ui

import cz.cuni.lf1.lge.ThunderSTORM.util.GridBagHelper
import cz.cuni.lf1.lge.ThunderSTORM.util.MacroUI.validators.DoubleValidatorFactory
import cz.cuni.lf1.lge.ThunderSTORM.util.MacroUI.validators.IntegerValidatorFactory
import cz.cuni.lf1.thunderstorm.algorithms.filters.CompoundWaveletFilter
import cz.cuni.lf1.thunderstorm.algorithms.filters.WaveletFilter
import cz.cuni.lf1.thunderstorm.algorithms.padding.DuplicatePadding
import cz.cuni.lf1.thunderstorm.algorithms.padding.ZeroPadding
import cz.cuni.lf1.thunderstorm.datastructures.GrayScaleImage
import java.awt.GridBagLayout
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JTextField

private const val name = "Wavelet filter (B-Spline)"

public class CompoundWaveletFilterUI : FilterUI() {

    private val order = parameters.createIntField("order", IntegerValidatorFactory.positiveNonZero(), 3)
    private val scale = parameters.createDoubleField("scale", DoubleValidatorFactory.positiveNonZero(), 2.0)

    public override fun getName()
            = name

    protected override fun getPreferencesPrefix()
            = super.getPreferencesPrefix() + ".wave"

    public override fun getOptionsPanel(): JPanel {
        val orderTextField = JTextField("", 20)
        val scaleTextField = JTextField("", 20)
        parameters.registerComponent(order, orderTextField)
        parameters.registerComponent(scale, scaleTextField)

        val panel = JPanel(GridBagLayout())
        panel.add(JLabel("B-Spline order: "), GridBagHelper.leftCol())
        panel.add(orderTextField, GridBagHelper.rightCol())
        panel.add(JLabel("B-Spline scale: "), GridBagHelper.leftCol())
        panel.add(scaleTextField, GridBagHelper.rightCol())
        parameters.loadPrefs()
        return panel
    }

    public override fun getImplementation()
            = CompoundWaveletFilter(
                order.value, scale.value,
                2*Math.ceil(order.value.toDouble()*scale.value/2.0).toInt() - 1,
                ::ZeroPadding)

    private fun getWaveletFilterImplementation(plane: Int)
            = WaveletFilter(
                plane, order.value, scale.value,
                2*Math.ceil(order.value.toDouble()*scale.value/2.0).toInt() - 1,
                ::DuplicatePadding)

    public override fun createThresholderSymbolTable(): Map<String, (GrayScaleImage) -> GrayScaleImage>
            = mapOf(
                Pair("Wave.F1", { img -> getWaveletFilterImplementation(1).filter(img) }),
                Pair("Wave.F2", { img -> getWaveletFilterImplementation(2).filter(img) }))
}